TM1py/Services/RESTService.py

```diff
@@ -148,40 +148,82 @@ def __exit__(self, exception_type, exception_value, traceback):
         self.logout()
 
     @httpmethod
-    def GET(self, request, data=''):
+    def GET(self, request, data='', headers=None, timeout=None):
         """ Perform a GET request against TM1 instance
-        :param request: String, for instance : /api/v1/Cubes?$top=1
-        :param data: String, empty
-        :return: String, the response as text
+        :param data: the payload
+        :param headers: custom headers
+        :param timeout: Number of seconds that the client will wait to receive the first byte.
+        :return: response object
         """
-        return self._s.get(url=request, headers=self._headers, data=data, verify=self._verify, timeout=self._timeout)
+        return self._s.get(
+            url=request,
+            headers={**self._headers, **headers} if headers else self._headers,
+            data=data,
+            verify=self._verify,
+            timeout=timeout if timeout else self._timeout)
 
     @httpmethod
-    def POST(self, request, data):
+    def POST(self, request, data, headers=None, timeout=None):
         """ POST request against the TM1 instance
-        :param request: String, /api/v1/Cubes
-        :param data: String, the payload (json)
-        :return:  String, the response as text
+        :param data: the payload
+        :param headers: custom headers
+        :param timeout: Number of seconds that the client will wait to receive the first byte.
+        :return: response object
         """
-        return self._s.post(url=request, headers=self._headers, data=data, verify=self._verify, timeout=self._timeout)
+        return self._s.post(
+            url=request,
+            headers={**self._headers, **headers} if headers else self._headers,
+            data=data,
+            verify=self._verify,
+            timeout=timeout if timeout else self._timeout)
 
     @httpmethod
-    def PATCH(self, request, data):
+    def PATCH(self, request, data, headers=None, timeout=None):
         """ PATCH request against the TM1 instance
         :param request: String, for instance : /api/v1/Dimensions('plan_business_unit')
-        :param data: String, the payload (json)
-        :return: String, the response as text
+        :param data: the payload
+        :param headers: custom headers
+        :param timeout: Number of seconds that the client will wait to receive the first byte.
+        :return: response object
         """
-        return self._s.patch(url=request, headers=self._headers, data=data, verify=self._verify, timeout=self._timeout)
+        return self._s.patch(
+            url=request,
+            headers={**self._headers, **headers} if headers else self._headers,
+            data=data,
+            verify=self._verify,
+            timeout=timeout if timeout else self._timeout)
```

Comment:
Hi @pbuncik,

please check if this implementation of setting custom headers on individual requests also serves your purpose.